File: colorlight.dart
// Send frames to a Colorlight 5A 75B receiver card
// It expects raw Ethernet frames.
// See https://hkubota.wordpress.com/2022/01/31/winter-project-colorlight-5a-75b-protocol/

// export nic=eth0 or whatever NIC you use
// Needs to run as root or via
// sudo setcap 'cap_net_admin,cap_net_raw+ep' sendframes.exe

import 'dart:ffi';
import 'dart:io';
import 'package:ffi/ffi.dart';
import 'package:l2ethernet/l2ethernet.dart';

// My LED matrix size

const columnCount = 128;
const rowCount = 64;

const IF_NAMESIZE = 16;

const frame0107DataLength = 98;
const frame0affDataLength = 63;
const frame5500DataLength = columnCount * 3 + 7;
final frameData0107 = calloc<Uint8>(frame0107DataLength);
final frameData0aff = calloc<Uint8>(frame0affDataLength);
final frameData5500 = calloc<Uint8>(frame5500DataLength);

// Brightness (max 255 for both)

const brightness = 0x28;
const brightnessPercent = 3;

void initFrames() {
  frameData0107[11] = brightnessPercent;
  frameData0107[12] = 5;
  frameData0107[14] = brightnessPercent;
  frameData0107[15] = brightnessPercent;
  frameData0107[16] = brightnessPercent;

  frameData0aff[0] = brightness;
  frameData0aff[1] = brightness;
  frameData0aff[2] = 255;

  frameData5500[0] = 0;
  frameData5500[1] = 0;
  frameData5500[2] = 0;
  frameData5500[3] = 0;
  frameData5500[4] = columnCount;
  frameData5500[5] = 0x08;
  frameData5500[6] = 0x88;
}

/// Calculate the next line in a pixel data frame (0x5500)
/// [row] is the row to calculate,
/// [frame] is the number of animation frame
void calculateFrame5500Row(int frame, int row) {
  for (int col = 0; col < columnCount; ++col) {
    if (col == frame || frame == row || row == 2 * rowCount - frame) {
      frameData5500[7 + 3 * col] = 255;
      frameData5500[7 + 3 * col + 1] = 255;
      frameData5500[7 + 3 * col + 2] = 255;
    } else {
      frameData5500[7 + 3 * col] = (64 - frame).abs();
      frameData5500[7 + 3 * col + 1] = frame;
      frameData5500[7 + 3 * col + 2] = (128 - frame).abs();
    }
  }
}

/// Cleanup the buffers for the Ethernet frames
void deleteFrames() {
  calloc.free(frameData0107);
  calloc.free(frameData0aff);
  calloc.free(frameData5500);
}

/// Open the raw socket, make a sweep sequence, repeat 10 times
void main() async {
  var ethName = Platform.environment["nic"];
  if (ethName == null) {
    print("Set nic environment variable first");
    exit(20);
  } else {
    var myl2eth = L2Ethernet(ethName);

    myl2eth.open();

    const src_mac = 0x222233445566;
    const dest_mac = 0x112233445566;
    const flags = 0;

    initFrames();

    // Draw as many frames as you have columns (128 in my case),
    // one vertical white line from left to right
    // to see tearing or lack of smoothness

    for (int k = 0; k < 10; ++k) {
      await sweep(myl2eth, src_mac, dest_mac, flags);
    }

    myl2eth.close();
    deleteFrames();
  }
}

const fps = 50;
const wait = true;
const waitTime = (1000 ~/ fps - 1);

/// Make one sweep sequence
///
Future<void> sweep(L2Ethernet l2, int src_mac, int dest_mac, int flags) async {
  int n;
  for (int t = 0; t < columnCount; ++t) {
    // Send a brightness packet

    n = l2.send(src_mac, dest_mac, 0x0a00 + brightness, frameData0aff,
        frame0affDataLength, flags);

    // Send one complete frame

    for (int y = 0; y < rowCount; ++y) {
      calculateFrame5500Row(t, y);
      frameData5500[0] = y;
      n = l2.send(
          src_mac, dest_mac, 0x5500, frameData5500, frame5500DataLength, flags);
    }

    // Without the following delay the end of the bottom row module flickers in the last line

    if (wait) await Future.delayed(Duration(milliseconds: 1));

    // Display frame

    n = l2.send(
        src_mac, dest_mac, 0x0107, frameData0107, frame0107DataLength, flags);

    // 20 fps, wait 50ms but subtract the 1ms from above

    if (wait) await Future.delayed(Duration(milliseconds: waitTime));
  }
}

File: node/src/lib/colorlight.ts
import {Eth} from './eth.ts';
import {Image} from 'imagescript'

// Show a horizontal and vertical sweeping line on the LED matrix connected to the Colorlight 5A 75B
export class ColorLight {

    width: number;
    height: number;
    brightnessPercent: number;
    brightnessValue: number;
    frame0107DataLength: number;
    frame0affDataLength: number;
    frame5500DataLength: number;
    frameData0107: Uint8Array;
    frameData0aff: Uint8Array;
    frameData5500: Uint8Array;
    eth: Eth;
    src_mac: bigint;
    dest_mac: bigint;
    flags: number;

    constructor(width: number, height: number, ethName: string) {
        this.width = width;
        this.height = height;

        this.brightnessPercent = 1;
        this.brightnessValue = 0x28;

        this.frame0107DataLength = 98;
        this.frame0affDataLength = 63;
        this.frame5500DataLength = this.width * 3 + 7;
        this.frameData0107 = new Uint8Array(this.frame0107DataLength).fill(0);
        this.frameData0aff = new Uint8Array(this.frame0affDataLength).fill(0);
        this.frameData5500 = new Uint8Array(this.frame5500DataLength).fill(0);
        this.initFrames();

        this.eth = new Eth(ethName);
        this.src_mac = 0x222233445566n;
        this.dest_mac = 0x112233445566n;
        this.flags = 0;
        this.eth.socketOpen();

    }

    set brightness(percent: number) {
        const brightnessMap = [
            [0, 0x00],
            [1, 0x03],
            [2, 0x05],
            [4, 0x0a],
            [5, 0x0d],
            [6, 0x0f],
            [10, 0x1a],
            [25, 0x40],
            [50, 0x80],
            [75, 0xbf],
            [100, 0xff]
        ];
        this.brightnessPercent = percent;
        for (let i = 0; i < brightnessMap.length; ++i) {
            const [p, v] = brightnessMap[i];
            if (percent >= p) {
                this.brightnessValue = v;
                this.initFrames();
                break;
            }
        }
    }


    private initFrames(): void {
        this.frameData0107[21] = this.brightnessPercent;
        this.frameData0107[22] = 5;
        this.frameData0107[24] = this.brightnessPercent;
        this.frameData0107[25] = this.brightnessPercent;
        this.frameData0107[26] = this.brightnessPercent;

        this.frameData0aff[0] = this.brightnessValue;
        this.frameData0aff[1] = this.brightnessValue;
        this.frameData0aff[2] = 255;

        this.frameData5500[0] = 0;
        this.frameData5500[1] = 0;
        this.frameData5500[2] = 0;
        this.frameData5500[3] = this.width >> 8;
        this.frameData5500[4] = this.width % 0xFF;
        this.frameData5500[5] = 0x08;
        this.frameData5500[6] = 0x88;
    }

    private frame5500fromImage(row: number, img: Image): void {
        this.frameData5500[0] = row;
        for (let col = 0, destOffset = 7, srcOffset = row * img.width * 4; col < this.width; ++col) {
            this.frameData5500[destOffset++] = img.bitmap[srcOffset + 2];
            this.frameData5500[destOffset++] = img.bitmap[srcOffset + 1];
            this.frameData5500[destOffset++] = img.bitmap[srcOffset];
            srcOffset += 4;
        }
    }

    async showImage(img: Image) {
        // Send a brightness packet
        let n = this.eth.send(this.src_mac, this.dest_mac, 0x0a00 + this.brightnessValue, this.frameData0aff,
            this.frame0affDataLength, this.flags);

        for (let t = 0; t < this.width; ++t) {
            // Send one complete frame

            for (let y = 0; y < this.height; ++y) {
                this.frame5500fromImage(y, img);
                this.frameData5500[0] = y;
                n = this.eth.send(this.src_mac, this.dest_mac, 0x5500, this.frameData5500, this.frame5500DataLength, this.flags);
            }

            // Without the following delay the end of the bottom row module flickers in the last line
            // if (wait) await Future.delayed(Duration(milliseconds: 1));
            await delay(1);

            n = this.eth.send(this.src_mac, this.dest_mac, 0x0107, this.frameData0107, this.frame0107DataLength, this.flags);
        }
    }
}

export async function delay(timeInMs: number): Promise<null> {
  await new Promise(resolve => setTimeout(() => {resolve()}, timeInMs))
}
